Parallel Compression improvements

jhendersonHDF opened the pull request.

jhendersonHDF committed:
Optimize 'array_gatherv' function and move it to H5mpi.c

jhendersonHDF committed:
Fix t_filters_parallel for MPI Communicator size of 1

jhendersonHDF committed:
Update parallel filters test

jhendersonHDF committed:
Parallel filters optimizations

Reduce memory usage of collective operations for parallel filtered
writes by sending only necessary info

Reduce memory usage of ranks sending chunk modification data by fixing
an incorrectly-sized malloc call

Eliminate some unnecessary chunk lookups by caching the information from
initial I/O list construction

jhendersonHDF committed:
Further memory usage reductions for parallel filtered writes

Free chunk data buffers before doing collective chunk reinsertions to
reduce memory usage

Create single MPI type for file and memory for Multi-Chunk I/O write
instead of two duplicate types

Remove a few now-unneeded arrays being allocated

jhendersonHDF committed:
Parallel filtered writes optimization

Add new H5D_select_io_mem routine to perform copies between memory
buffers according to selections in two different dataspaces using
H5VM_memcpyvv

Use new H5D_select_io_mem routine to avoid unnecessary data copies
during parallel filtered writes

jhendersonHDF committed:
Parallel filtered writes collective I/O updates

Switch to actually using collective I/O for chunk reads during
linked-chunk and multi-chunk reads/writes. Previously, only chunk write
operations were done collectively

Enable linked-chunk I/O for collective filtered reads

jhendersonHDF committed:
Update MPI I/O driver debugging to print memory type for reads/writes

jhendersonHDF committed:
Add H5D package routine to retrieve strings for the reasons for breaking collective I/O

jhendersonHDF committed:
Parallel Filters optimization

Reduce memory usage by MPI rank 0 during chunk redistribution

jhendersonHDF committed:
Add initial H5Dmpio-level debugging

jhendersonHDF committed:
Cleanup derived datatype creation for parallel compression

jhendersonHDF committed:
Add H5pubconf.h #ifdef for parallel filtered writes support

jhendersonHDF committed:
Add examples for parallel compression feature

jhendersonHDF committed:
Parallel compression memory usage reduction

Heavily reduce memory usage for parallel compression by calculating
chunk 'scaled' coordinates on the fly, rather than communicating them
between ranks

jhendersonHDF committed:
Sort by secondary key in H5D__cmp_filtered_collective_io_info_entry routine

jhendersonHDF commented on src/H5Dio.c:
```diff
@@ -810,98 +810,35 @@ H5D__ioinfo_adjust(H5D_io_info_t *io_info, const H5D_t *dset, const H5S_t *file_
+                    char global_no_coll_cause_string[512];
+
+                    if (H5D__mpio_get_no_coll_cause_strings(local_no_coll_cause_string, 512,
+                                                            global_no_coll_cause_string, 512) < 0)
```
Moved most of this code into a new function to get strings for the reasons why collective I/O was broken for code reuse.

jhendersonHDF commented on src/H5Dselect.c:
```diff
@@ -256,6 +256,187 @@ H5D__select_io(const H5D_io_info_t *io_info, size_t elmt_size, size_t nelmts, co
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5D_select_io_mem(void *dst_buf, const H5S_t *dst_space, const void *src_buf, const H5S_t *src_space,
```
A new routine that is very similar to H5D__select_io(), but rather than copying between application memory and the file, copies between two memory buffers according to the selection in the dst and src dataspaces.

jhendersonHDF commented on src/H5FDmpio.c:
```diff
@@ -188,6 +188,41 @@ H5FD__mpio_parse_debug_str(const char *s)
+ *---------------------------------------------------------------------------
+ */
+static const char *
+H5FD__mem_t_to_str(H5FD_mem_t mem_type)
```
The changes in this file allow one to see what type of I/O the MPI I/O file driver is doing. Previously one would only see the offset and length of the I/O. This now also shows you whether it's a superblock area, raw data, object header, etc.

jhendersonHDF commented on src/H5mpi.c:
```diff
@@ -549,4 +549,238 @@ H5_mpio_create_large_type(hsize_t num_elements, MPI_Aint stride_bytes, MPI_Datat
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5_mpio_gatherv_alloc(void *send_buf, int send_count, MPI_Datatype send_type, const int recv_counts[],
```
The two new functions here are simply wrappers around MPI_(All)gatherv that hide a bit of boilerplate code. Both allocate the receive buffer for the caller. The only difference between the two is that the "simple" function calculates the recv_counts and displacements arrays for the caller before making the MPI_(All)gatherv call.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -273,6 +373,185 @@ static int H5D__cmp_filtered_collective_io_info_entry_owner(const void *filtered
 /* Local Variables */
 /*******************/
 
+#ifdef H5Dmpio_DEBUG
```
The below code adds debugging to H5Dmpio similar to that in the MPI I/O file driver.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -499,145 +892,6 @@ H5D__mpio_select_write(const H5D_io_info_t *io_info, const H5D_type_info_t H5_AT
- *-------------------------------------------------------------------------
- */
-static herr_t
-H5D__mpio_array_gatherv(void *local_array, size_t local_array_num_entries, size_t array_entry_size,
```
This whole routine was rewritten and move to H5mpi.c

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -808,6 +1068,36 @@ H5D__chunk_collective_io(H5D_io_info_t *io_info, const H5D_type_info_t *type_inf
+    if ((mpi_rank = H5F_mpi_get_rank(io_info->dset->oloc.file)) < 0)
+        HGOTO_ERROR(H5E_IO, H5E_MPI, FAIL, "unable to obtain MPI rank")
+    if ((mpi_size = H5F_mpi_get_size(io_info->dset->oloc.file)) < 0)
+        HGOTO_ERROR(H5E_IO, H5E_MPI, FAIL, "unable to obtain MPI size")
```
Rather than retrieving the MPI rank and size multiple times in this file, do it once in H5D__chunk_collective_io which tends to be the main entrypoint in this file. Then, just hand those down to functions as needed.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -1376,124 +1666,126 @@ H5D__link_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_in
+     * without reading metadata collectively and it should be safe to enable
+     * them here because it is required that these dataset writes be collective.
+     */
+    H5CX_set_coll_metadata_read(TRUE);
```
The changes here currently turn on collective metadata reads for the feature since it will not scale well without them and we require the writes to be collective, but this may or may not be the best idea depending on the number of chunks in the operation.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -1376,124 +1666,126 @@ H5D__link_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_in
+        /* Redistribute any shared chunks being written to and
+         * send around chunk modification data as necessary
          */
-        if (H5D__mpio_array_gatherv(chunk_list, chunk_list_num_entries,
```
Rather than gathering everybody's list of chunks into a collective array, the feature has been revised in most places to construct different MPI derived types to only send as much data as needed, greatly reducing the feature's memory usage.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -2954,464 +3400,1539 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
  *-------------------------------------------------------------------------
  */
 static herr_t
-H5D__filtered_collective_chunk_entry_io(H5D_filtered_collective_io_info_t *chunk_entry,
```
This routine used to work on either reading an individual chunk (for dataset reads) or reading and writing an individual chunk (for dataset writes). However, any chunk reads here used to be independent which is a scalability problem for the feature. The new H5D__mpio_collective_filtered_chunk_read, H5D__mpio_collective_filtered_chunk_update and H5D__mpio_collective_filtered_chunk_common_io routines now perform the duties of this routine, but in a manner that allows chunk reads to be done collectively. This should generally scale much better and still allows the user the option of specifying independent chunk reads when desired.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -2954,464 +3400,1539 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+} /* end H5D__mpio_collective_filtered_chunk_reinsert() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5D__mpio_get_chunk_redistribute_info_types
```
The 3 functions below here create different MPI derived datatypes to extract certain portions of information from the overall per-chunk H5D_filtered_collective_io_info_t structure. Usually, a particular operation (shared chunk redistribution, chunk reallocation, chunk reinsertion) only needs a few fields out of that structure and this information is gathered to all ranks, so sending just the few fields necessary can drastically save on memory usage at the expense of a bit of MPI overhead.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -2954,464 +3400,1539 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5D__mpio_collective_filtered_io_type(H5D_filtered_collective_io_info_t *chunk_list, size_t num_entries,
```
This routine was just revised a little bit to create slightly more efficient MPI derived types for performing I/O on filtered chunks.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -1815,202 +2104,173 @@ H5D__multi_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_i
             /* Participate in the collective re-insertion of all chunks modified
              * in this iteration into the chunk index
              */
-            for (j = 0; j < collective_chunk_list_num_entries; j++) {
```
Chunk index reinsertion logic here moved into H5D__mpio_collective_filtered_chunk_reinsert, which more efficiently handles memory usage as well as chunk reinsertion itself.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -1376,124 +1666,126 @@ H5D__link_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_in
-         * in this iteration into the chunk index
+         * into the chunk index
          */
-        for (i = 0; i < collective_chunk_list_num_entries; i++) {
```
Chunk index reinsertion logic here moved into H5D__mpio_collective_filtered_chunk_reinsert, which more efficiently handles memory usage as well as chunk reinsertion itself.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -1815,202 +2104,173 @@ H5D__multi_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_i
+             * use MPI_BYTE for the file and memory type and specify
+             * a count of 0.
              */
-            for (j = 0; j < collective_chunk_list_num_entries; j++) {
```
Chunk file space reallocation logic moved into H5D__mpio_collective_filtered_chunk_reallocate, which more efficiently handles memory usage.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -1376,124 +1666,126 @@ H5D__link_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_in
-            HGOTO_ERROR(H5E_DATASET, H5E_CANTGATHER, FAIL, "couldn't gather new chunk sizes")
-
-        /* Collectively re-allocate the modified chunks (from each process) in the file */
-        for (i = 0; i < collective_chunk_list_num_entries; i++) {
```
Chunk file space reallocation logic moved into H5D__mpio_collective_filtered_chunk_reallocate, which more efficiently handles memory usage.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -1815,202 +2104,173 @@ H5D__multi_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_i
+            have_chunk_to_process = (i < chunk_list_num_entries) && (mpi_rank == chunk_list[i].new_owner);
 
-            if (have_chunk_to_process)
-                if (H5D__filtered_collective_chunk_entry_io(&chunk_list[i], io_info, type_info, fm) < 0)
```
Duties now performed by H5D__mpio_collective_filtered_chunk_update instead.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -1376,124 +1666,126 @@ H5D__link_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_in
-         */
-        for (i = 0; i < chunk_list_num_entries; i++)
-            if (mpi_rank == chunk_list[i].owners.new_owner)
-                if (H5D__filtered_collective_chunk_entry_io(&chunk_list[i], io_info, type_info, fm) < 0)
```
Duties now performed by H5D__mpio_collective_filtered_chunk_update instead.

jhendersonHDF commented on src/H5Dmpio.c:
```diff
@@ -1815,202 +2104,173 @@ H5D__multi_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_i
 
     if (io_info->op_type == H5D_IO_OP_READ) { /* Filtered collective read */
-        for (i = 0; i < chunk_list_num_entries; i++)
-            if (H5D__filtered_collective_chunk_entry_io(&chunk_list[i], io_info, type_info, fm) < 0)
```
Duties now performed by H5D__mpio_collective_filtered_chunk_read instead.

jhendersonHDF committed:
Set correct pixels-per-block option for SZIP in t_filters_parallel test

lrknox commented:
These tests failed with module  OpenMPI/4.0.2-GCC-7.2.0-2.29 on jelly:
t_bigio
testphdf5 -- compressed dataset collective read (cmpdsetr)
Paralle Filters test - Testing write to unshared filtered chunks
t_2Gio -- compressed dataset collective read (cmpdsetr)

Error messages:
t_bigio - 

Single Rank Independent I/O
HDF5-DIAG: Error detected in HDF5 (1.13.0-7) MPI-process 0:
  #000: ../../src/H5D.c line 1222 in H5Dwrite(): can't synchronously write data
    major: Dataset
    minor: Write failed
  #001: ../../src/H5D.c line 1169 in H5D__write_api_common(): can't write data
    major: Dataset
    minor: Write failed
  #002: ../../src/H5VLcallback.c line 2181 in H5VL_dataset_write(): dataset write failed
    major: Virtual Object Layer
    minor: Write failed
  #003: ../../src/H5VLcallback.c line 2148 in H5VL__dataset_write(): dataset write failed
    major: Virtual Object Layer
    minor: Write failed
  #004: ../../src/H5VLnative_dataset.c line 345 in H5VL__native_dataset_write(): can't write data
    major: Dataset
    minor: Write failed
  #005: ../../src/H5Dio.c line 475 in H5D__write(): can't write data
    major: Dataset
    minor: Write failed
  #006: ../../src/H5Dcontig.c line 628 in H5D__contig_write(): contiguous write failed
    major: Dataset
    minor: Write failed
  #007: ../../src/H5Dselect.c line 492 in H5D__select_write(): write error
    major: Dataspace
    minor: Write failed
  #008: ../../src/H5Dselect.c line 224 in H5D__select_io(): write error
    major: Dataspace
    minor: Write failed
  #009: ../../src/H5Dcontig.c line 1249 in H5D__contig_writevv(): can't perform vectorized read
    major: Dataset
    minor: Can't operate on object
  #010: ../../src/H5VM.c line 1401 in H5VM_opvv(): can't perform operation
    major: Internal error (too specific to document in detail)
    minor: Can't operate on object
  #011: ../../src/H5Dcontig.c line 1179 in H5D__contig_writevv_cb(): block write failed
    major: Dataset
    minor: Write failed
  #012: ../../src/H5Fio.c line 191 in H5F_shared_block_write(): write through page buffer failed
    major: Low-level I/O
    minor: Write failed
  #013: ../../src/H5PB.c line 1018 in H5PB_write(): write through metadata accumulator failed
    major: Page Buffering
    minor: Write failed
  #014: ../../src/H5Faccum.c line 831 in H5F__accum_write(): file write failed
    major: Low-level I/O
    minor: Write failed
  #015: ../../src/H5FDint.c line 256 in H5FD_write(): driver write request failed
    major: Virtual File Layer
    minor: Write failed
  #016: ../../src/H5FDmpio.c line 1597 in H5FD__mpio_write(): file write failed
    major: Low-level I/O
    minor: Write failed

Test log for testphdf5
...
Testing  -- compressed dataset collective read (cmpdsetr)
[jelly:08172] *** Process received signal ***
[jelly:08172] Signal: Segmentation fault (11)
[jelly:08172] Signal code: Address not mapped (1)
[jelly:08172] Failing at address: 0x7ffd6aa548d0
[jelly:08172] [ 0] /lib64/libc.so.6(+0x36450)[0x7fc9f800a450]
[jelly:08172] [ 1] /lib64/libc.so.6(+0x154d0b)[0x7fc9f8128d0b]
...

Test log for t_filters_parallel
...
Testing write to unshared filtered chunks
[jelly:42035] *** Process received signal ***
[jelly:42035] Signal: Segmentation fault (11)
[jelly:42035] Signal code: Address not mapped (1)
[jelly:42035] Failing at address: 0x7ffe4e575000

Test log for t_2Gio
...
Testing  -- compressed dataset collective read (cmpdsetr)
[jelly:42526] *** Process received signal ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec noticed that process rank 0 with PID 0 on node jelly exited on signal 11 (Segmentation fault).

ph5diff was also did hang, but I think that may be unrelated to this PR.

The complete output is in/mnt/wrk/lrknox/jhenderson-hdf5/atbuild/makecheckp.out for now.

lrknox commented:
Tests passed with MPICH 3.3 and with OpenMPI 4.1.

derobins commented on examples/ph5_filtered_writes.c:
```diff
@@ -0,0 +1,478 @@
+         * Set 'deflate' filter with reasonable
+         * compression level on DCPL
+         */
+        H5Pset_deflate(dcpl_id, 6);
```
If H5Zfilter_avail() fails, the FAIL (-1) value will be interpreted as truthy and this code will be inappropriately called. The other example also exhibits this behavior.

jhendersonHDF replied:
good catch, will fix

derobins commented on src/H5public.h:
```diff
@@ -324,6 +324,23 @@ typedef uint64_t haddr_t;
+#else
+#error "no suitable MPI type found for hsize_t"
+#endif
+
```
I don't think we need this level of complexity. All 64-bit data models define long long as 64 bits, which is what we've hard-coded hsize_t to be.

I would:
- Define the type to MPI_UNSIGNED_LONG_LONG
- Either name the hsize_t type HSIZE_AS_MPI_TYPE or rename the haddr_t type H5_MPI_HADDR_T so the types have name parity
- Move the values closer to the definitions of hsize_t and haddr_t, respectively so they are less likely to get out of sync

jhendersonHDF replied:
I agree with most all of this. The only thing I would mention is that having the catch for not finding a suitable type to use for the definition is a bit more future proof than just defining the type as MPI_UNSIGNED_LONG_LONG. Though, whether or not haddr_t/hsize_t will ever be larger than 64 bits remains to be seen. The comparisons against sizes < 64 bits are probably unnecessary here however.

jhendersonHDF committed:
Fix parallel filters examples for case where H5Zfilter_avail fails

jhendersonHDF committed:
Use free lists for dataspace selection iterators in parallel compression feature

jhendersonHDF committed:
Don't force collective metadata reads for parallel compression feature

jhendersonHDF committed:
Fix parallel compression for extensible arrays

jhendersonHDF committed:
Minor parallel compression updates

jhendersonHDF committed:
Add tests for parallel compression with extensible arrays and btrees

jhendersonHDF committed:
Simplify MPI type definitions for haddr_t and hsize_t

jhendersonHDF committed:
Fix parallel examples to share common TEST_REFERENCE

derobins commented on testpar/t_filters_parallel.c:
```diff
@@ -5554,16 +5379,23 @@ test_read_cmpd_filtered_dataset_type_conversion_unshared(void)
+    hid_t                file_id = H5I_INVALID_HID, dset_id = H5I_INVALID_HID, plist_id = H5I_INVALID_HID;
+    hid_t                filetype = H5I_INVALID_HID, memtype = H5I_INVALID_HID;
+    hid_t                group_id  = H5I_INVALID_HID;
+    hid_t                filespace = H5I_INVALID_HID, memspace = H5I_INVALID_HID;
```
I usually put these one per line. I find them hard to parse when glommed together with commas.

jhendersonHDF replied:
I mostly tried to just put related fields on the same line so that the function's declarations didn't get excessively long vertically, but I can always change this if it's hard to parse. With the introduction of H5S_BLOCK, I eliminated some old cruft variables, but didn't go over everything.

derobins commented on testpar/t_filters_parallel.c:
```diff
@@ -774,15 +808,17 @@ test_write_filtered_dataset_single_no_selection(void)
-    data = (C_DATATYPE *)HDcalloc(1, data_size);
-    VRFY((NULL != data), "HDcalloc succeeded");
+    if (mpi_rank != WRITE_SINGLE_NO_SELECTION_FILTERED_CHUNKS_NO_SELECT_PROC) {
+        data = (C_DATATYPE *)HDcalloc(1, data_size);
```
Not a huge deal, but there's no need to cast the results of allocations.

jhendersonHDF replied:
Very true. Just copy and pasted from old code that was written when I was less familiar with C conventions.

derobins commented on src/H5mpi.c:
```diff
@@ -549,4 +549,238 @@ H5_mpio_create_large_type(hsize_t num_elements, MPI_Aint stride_bytes, MPI_Datat
+
+done:
+    if (ret_value < 0) {
+        if (recv_buf)
```
Freeing NULL pointers is accepted by the standard so the check is superfluous.

jhendersonHDF replied:
Indeed. Switching to just freeing the pointer is something I've only started doing sort of recently but we should probably be doing that everywhere in the library.

derobins commented on src/H5Dselect.c:
```diff
@@ -256,6 +256,187 @@ H5D__select_io(const H5D_io_info_t *io_info, size_t elmt_size, size_t nelmts, co
+    size_t          src_nseq;                  /* Number of sequences generated for source buffer */
+    size_t          dxpl_vec_size;             /* Vector length from API context's DXPL */
+    size_t          vec_size;                  /* Vector length */
+    ssize_t         bytes_copied;
```
Not critical for this PR, but we should consider refactoring H5VM_memcpyvv to return herr_t and return the number of bytes copied via an out parameter.

jhendersonHDF replied:
Agreed. A memcpy implementation is certainly one where I could see the range of ssize_t being a problem one day. I'll make a note on changing this.

jhendersonHDF committed:
Implement support for 'don't filter partial edge chunks' flag in parallel compression

jhendersonHDF committed:
Remove old #ifdef'ed out parallel compression test code

jhendersonHDF committed:
Fix H5Dget_space_status for filtered datasets

jhendersonHDF committed:
Parallel compression feature support

Add support for chunk fill values to parallel compression feature

Add partial support for incremental file space allocation to parallel compression feature

jhendersonHDF committed:
Implement full support for incremental file space allocation with parallel compression

jhendersonHDF committed:
Update chunk redistribution to be more efficient

jhendersonHDF committed:
Parallel compression MPI overhead reduction

Refactor chunk reallocation and reinsertion code to use less MPI communication during linked-chunk I/O

fortnern commented on src/H5Dselect.c:
```diff
@@ -256,6 +256,187 @@ H5D__select_io(const H5D_io_info_t *io_info, size_t elmt_size, size_t nelmts, co
+            } /* end if */
+
+            /* Perform vectorized memcpy from src_buf to dst_buf */
+            if ((bytes_copied = H5VM_memcpyvv(dst_buf, dst_nseq, &curr_dst_seq, dst_len, dst_off, src_buf,
```
Are we sue this algorithm correctly handles the case where src and dst have different numbers of elements? This could happen get_seq_list is terminated due to vec_size being too small.

fortnern replied:
Nevermind, I think this algorithm is fine.

jhendersonHDF committed:
Update chunk modification data sharing to use a nonblocking consensus algorithm

fortnern commented on src/H5mpi.c:
```diff
@@ -549,4 +549,238 @@ H5_mpio_create_large_type(hsize_t num_elements, MPI_Aint stride_bytes, MPI_Datat
+        size_t buf_size;
+
+#if MPI_VERSION >= 3
+        H5_CHECK_OVERFLOW(type_extent, MPI_Count, size_t);
```
There should probably be a comment explaining why is this overflow check only done for MPI version > 3. Is it because lower versions will never have MPI_Aint larger than size_t?

Also: consider changing int to MPI_Aint for MPI version 2 routines where appropriate

jhendersonHDF replied:
Removed in latest changes as I now check for a negative extent first and casting should be safe for positive values.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -436,6 +714,125 @@ H5D__mpio_opt_possible(const H5D_io_info_t *io_info, const H5S_t *file_space, co
+    HDassert((local_cause && local_cause_len > 0) || (global_cause && global_cause_len > 0));
+
+    /* Ensure things are in order for our cause string table */
+    HDcompile_assert(H5D_MPIO_SET_INDEPENDENT == (H5D_mpio_no_collective_cause_t)1);
```
Minor nitpick: I like to use hexadecimal for literal bitmap values like this

jhendersonHDF replied:
Updated this routine to be a bit cleaner about the enum/bit value handling.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -1377,123 +1671,118 @@ H5D__link_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_in
+             * Override the write buffer to point to the first
+             * chunk's data buffer
              */
             io_info->u.wbuf = chunk_list[0].buf;
```
Interesting - so this works even if the buffers aren't in the same order as the chunks? So some buffers have a negative offset (before factoring in this base address)?

jhendersonHDF replied:
Yes, the type created for the chunk memory buffers tends to have scattered buffers and the negative offsets for these are perfectly fine with MPI. We just have to make sure the chunks are sorted by file address.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -2675,276 +2975,420 @@ H5D__construct_filtered_io_info_list(const H5D_io_info_t *io_info, const H5D_typ
-            num_assigned_chunks_array[new_chunk_owner]++;
-        } /* end for */
+                    chunk_entry++;
+                } while (++i < all_chunks_array_num_entries &&
```
Should probably add a comment here or at the for loop about how the "i" variable is advanced both by the for loop and this loop, to improve readability

jhendersonHDF replied:
Added comments for this.

soumagne commented on src/H5Dmpio.c:
```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+     * structure. Then, resize the type with an extent equal to the size of the
+     * encompassing structure to make it suitable for sending arrays of them.
+     */
+    {
```
was there a reason for making a separate block?

jhendersonHDF replied:
In this case, it was just to have a smaller scope for the 'extent' variable that we throw away after the call to MPI_Type_get_extent. To create the resized type, I only needed the lower bound ('lb') for the original type and not the extent, but I believe MPI_Type_get_extent requires you to provide variables for both 'lb' and 'extent'.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
-                    if (MPI_SUCCESS != (mpi_code = MPI_Mprobe(MPI_ANY_SOURCE, (int)chunk_entry->index,
-                                                              io_info->comm, &message, &status)))
+                    if (MPI_SUCCESS !=
+                        (mpi_code = MPI_Mprobe(MPI_ANY_SOURCE, (int)chunk_entry->index_info.chunk_idx,
```
Seems strange to use a blocking call here then a non-blocking irecv, though I guess it simplifies the code. Do you think there would be performance to be gained by using MPI_Improbe (and adding another pass over the chunks)?

jhendersonHDF replied:
Here I think the changes are mostly just formatting and struct field renaming, so this blocking probe approach was the old first attempt at getting this correct. This has since been re-written into a non-blocking probing approach (not in this current PR), but I think in general it can be hard to guess about performance since we might be entering MPI multiple times in a non-blocking probing loop as opposed to blocking here. Hard to say, but I think initial profiling with the newly-rewritten non-blocking way will give some ideas.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
                     if (MPI_SUCCESS != (mpi_code = MPI_Get_count(&status, MPI_BYTE, &count)))
                         HMPI_GOTO_ERROR(FAIL, "MPI_Get_count failed", mpi_code)
 
                     HDassert(count >= 0);
```
I would prefer a check and error here instead of an assertion, since assertions are supposed to only be used to guarantee a code's internal consistency, and this depends on MPI and the network. You could make an argument otherwise though

jhendersonHDF replied:
So, I agree with this point and did change it to an error check instead of an assert in my current set of changes. That said, I think the only danger here is if the MPI implementation is buggy/broken, in which case we probably have bigger problems. But yes, an assert probably wasn't best to use here, so I switched it (final changes to come in another PR rather than making this one larger).

fortnern commented on src/H5Dmpio.c:
```diff
@@ -1816,201 +2102,165 @@ H5D__multi_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_i
+                HGOTO_ERROR(H5E_DATASET, H5E_WRITEERROR, FAIL, "unable to redistribute shared chunks")
+
+            /* Send chunk modification data */
+            if (H5D__mpio_share_chunk_modification_data(chunk_list, &chunk_list_num_entries, io_info,
```
Doesn't sharing all modification data at once use (potentially) lots of memory? If the primary purpose of multi chunk is to save memory, should modification data sharing be done in the loop, once per chunk? Or would the increased communication outweigh the memory savings?

jhendersonHDF replied:
Yes, this is generally where a lot of memory usage will probably be encountered. However, for multi-chunk things get a little bit complicated because a rank can't finish updating a chunk until it's received all messages for that chunk. So we could end up in a situation where ranks are sending messages to other ranks for chunks that those ranks aren't processing yet. Things could be refactored a bit so that everyone keeps a collective list of all the chunks sorted in the same order and then all ranks process in order, either processing the chunk if the rank owns it, or sending a message out if not. That would probably increase memory usage a decent bit though (though probably not as much as the modification message overhead), and would probably have to use blocking communications since chunks are done one at a time. Not the end of the world or anything, but there's certainly room for overlapping non-blocking communications with some parallel compression work.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+                    *local_entry = *coll_entry;
+                } while ((++num_processed < chunk_list_num_entries) && (++i < collective_num_entries));
+
+                continue;
```
Couldn't you just use an "else" instead of this continue?

jhendersonHDF replied:
Yes, the `continue` was more to just make this slightly odd section of code stand out, since it might be non-obvious what it's doing. If you'd prefer a simpler `else` for readability, I can change that.

jhendersonHDF replied:
Made much cleaner in latest changes.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+         * of the modified chunks into the chunk index
+         */
+        H5_CHECK_OVERFLOW(chunk_list_num_entries, size_t, int);
+        if (H5_mpio_gatherv_alloc_simple(chunk_list, (int)chunk_list_num_entries, send_type, recv_type, TRUE,
```
Would it be possible to avoid this allgather if we kept the needed data around for all chunks? I read the comment above saying it wasn't necessary to keep need_insert, but is there another reason we need to do the allgather?

jhendersonHDF replied:
Since the struct fields used for the two operations (chunk file space allocation and chunk index re-insertion) differ from each other slightly, two allgathers are used mostly to save on memory. E.g., allgather only the info needed for re-allocation -> re-allocate -> free collective array, then allgather only the info needed for re-insertion -> re-insert -> free collective array. The `chunk_idx` and `chunk_new` fields are shared between the two operations, but the `chunk_current`, `filter_mask` and `need_insert` fields aren't.

It might seem like a bit of a micro-optimization, but would use around 24-32 bytes extra per chunk, multiplied by the number of ranks that have a copy of these collective arrays, so it can add up. I'm hoping that the memory saving is more useful than the overhead of the collective operations, but if two allgathers is prohibitive, the collective fields needed could be combined again.

fortnern replied:
I noticed you changed the code in realloc to always save need_insert, but it seems these allgathers are still here?

fortnern commented on src/H5Dmpio.c:
```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+                }
+                else {
+                    H5_CHECK_OVERFLOW(chunk_list[i].fspace_info.chunk_new.length, hsize_t, int);
+                    length_array[chunk_count] = (int)chunk_list[i].fspace_info.chunk_new.length;
```
It might be worth setting up a convenience pointer to chunk_current/chunk_new to avoid these read/write branches and make the code a little more readable.

jhendersonHDF replied:
I think that's a good idea, especially for readability.

jhendersonHDF replied:
Fixed.

jhendersonHDF committed:
Use paged persistent free space management in parallel compression tests

jhendersonHDF committed:
Cleanup from initial parallel compression review

github-actions[bot] committed:
Committing clang-format changes

fortnern commented on src/H5Dmpio.c:
```diff
@@ -2997,6 +3003,7 @@ H5D__mpio_collective_filtered_chunk_io_setup(const H5D_io_info_t *io_info, const
                                              size_t *num_entries, int mpi_rank)
 {
     H5D_filtered_collective_io_info_t *local_info_array = NULL;
+    hbool_t                            filter_partial_edge_chunks;
```
Seems this is the inverse of how the option is usually presented. Not a big deal, but could cause some confusion.

jhendersonHDF replied:
Just seemed more natural to think of it this way for me. Perfectly happy to change/rename it if you think it'll be confusing.

fortnern commented on src/H5Dint.c:
```diff
@@ -384,37 +384,60 @@ H5D__get_space_status(const H5D_t *dset, H5D_space_status_t *allocation)
+                HGOTO_ERROR(H5E_DATASET, H5E_OVERFLOW, FAIL, "size of dataset's storage overflowed")
+
+            /* Difficult to error check, since the error value is 0 and 0 is a valid value... :-/ */
+            if (H5D__get_storage_size(dset, &space_allocated) < 0)
```
Is this better than H5D__get_num_chunks() for the unfiltered case or could we just always use H5D__get_num_chunks()?

jhendersonHDF replied:
It looks like we could probably just use H5D__get_num_chunks here; I'll need to test that to make sure.

soumagne reviewed (approved):
Approving assuming performance is validated. Should not be merged before.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -3135,96 +3172,213 @@ H5D__mpio_collective_filtered_chunk_io_setup(const H5D_io_info_t *io_info, const
+        HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL, "couldn't allocate assigned chunks array")
+
+    /* Perform initial Allgather to determine the collective chunk list size */
+    if (MPI_SUCCESS != (mpi_code = MPI_Allgather(&chunk_list_num_entries, 1, H5_SIZE_T_AS_MPI_TYPE,
```
As discussed in the meeting, I'm not convinced there's a substantial reason to use anything other than p0. Is it really likely that choosing a different algorithm here could save more time than this allgather necessary to make the choice adds?

jhendersonHDF replied:
Keep in mind that an initial (all)gather is necessary anyway because someone or everyone needs to know how big of a collective array to allocate for redistribution, so it's not really adding anything in terms of overhead. The allgather here also creates the rank -> num chunks assigned map which saves us two allgather operations later on in the reallocate and reinsert operations. Those used to each do 2 operations - first an allgather to see how many chunks everyone's contributing to the collective array (all of them for linked-chunk I/O, or possibly just 0 or 1 per rank for multi-chunk I/O) and then an allgatherv to actually gather the chunks. The rank -> num chunks assigned map lets us skip the initial allgather in each and just do the allgatherv, so we knock off 2 collective operations.

And remember that this is more about memory usage than performance, since memory usage multiplies by the number of ranks that each have to keep a copy of these distributed arrays. In any case, I feel like this is something that just needs to be tested at scale to see if this optimization is relevant or not. I think the next step is really to try using MPI-3's one-sided RDMA-like operations rather than dealing with all these collective arrays.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -3321,18 +3489,25 @@ H5D__mpio_redistribute_shared_chunks_p0(H5D_filtered_collective_io_info_t *local
         HMPI_GOTO_ERROR(FAIL, "unable to scatter shared chunks info buffer", mpi_code)
 
+    /* Distribute rank number -> assigned chunks mapping back to all ranks */
+    if (MPI_SUCCESS != (mpi_code = MPI_Bcast(num_chunks_assigned_map, mpi_size, H5_SIZE_T_AS_MPI_TYPE,
```
How did this work without the bcast before?

jhendersonHDF replied:
Note that the rank -> assigned chunks mapping is newly-added with the recent round of commits, so the MPI_Bcast goes along with that.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -3683,105 +3820,204 @@ H5D__mpio_share_chunk_modification_data(H5D_filtered_collective_io_info_t *chunk
-    for (i = 0, last_assigned_idx = 0; i < *chunk_list_num_entries; i++) {
-        H5D_filtered_collective_io_info_t *chunk_entry = &chunk_list[i];
+    /* Check if the number of send or receive requests will overflow an int (MPI requirement) */
+    H5_CHECK_OVERFLOW(num_send_requests, size_t, int)
```
Hmm is it possible to print a more informative error message here? Maybe unlikely that it'll be reasonable to post this many sends in the near future but it'd at least give app developers a hint that their I/O setup is bad

jhendersonHDF replied:
I think that might be useful. Certainly if parallel compression is used at exascale (or maybe even smaller) we could probably start sending too many requests out and will have to rethink a little bit about how to break down the problem size. Let me convert this to something more helpful.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -4228,135 +4462,154 @@ H5D__mpio_collective_filtered_chunk_update(H5D_filtered_collective_io_info_t *ch
+        /*
+         * If this chunk hasn't been allocated yet and we aren't writing
+         * out fill values to it, make sure to 0-fill its memory buffer
+         * so we don't use uninitialized memory.
```
This behavior is controverial, see recent discussions with Suren. Though in the case of compression using malloc would likely do more harm than good.

jhendersonHDF replied:
I agree on the controversial part. However, I was previously using malloc and of course seeing lots of warnings about conditional jumps based on uninitialized values inside different filters' logic. While likely harmless there, 0-filling here hopefully shouldn't be as expensive as actually filling the memory buffer with a fill value and cleans up false positives. So, to be consistent with H5D__chunk_lock, I just chose to do the same thing as there and 0-fill the buffer (Calloc here, memset there). Users may expect to get junk, but it seems good to play it safe for the filters if the users don't care in the first place.

fortnern commented on src/H5Dmpio.c:
```diff
@@ -3620,55 +3717,95 @@ H5D__mpio_share_chunk_modification_data(H5D_filtered_collective_io_info_t *chunk
+             * message itself will contain the chunk's index so we can update
+             * the correct chunk with the received data.
+             */
+            HASH_ADD(hh, chunk_table, index_info.chunk_idx, sizeof(hsize_t), &chunk_list[last_assigned_idx]);
```
I'm having trouble understanding this. It seems like you're using an hsize_t as the index into the hash table, which might be larger than an int.

jhendersonHDF replied:
Yes, uthash appears to use an unsigned for the hash value, so there are going to be hash collisions here. It's maybe not ideal, but since the chunk index can't be used for the MPI message tag and messages can come in any order, I needed a different way to quickly find a particular chunk structure by chunk index without iterating through the local array each time.

If we were to convert the library's chunk info structure to also include stuff needed for parallel compression, we could probably just do a skip list search to find a chunk. That may or may not be better here on average depending on the number of hash collisions and how well collisions are resolved, but there's a decent bit more work to do there. Open to suggestions.

fortnern commented on src/H5Dselect.c:
```diff
@@ -256,6 +256,187 @@ H5D__select_io(const H5D_io_info_t *io_info, size_t elmt_size, size_t nelmts, H5
+
+            /* Decrement number of elements left to process */
+            HDassert(((size_t)bytes_copied % elmt_size) == 0);
+            if (elmt_size == 0)
```
Shouldn't this just be an assert at the top of the function?

jhendersonHDF replied:
Hmm, this was copied from H5D__select_io, but yes, I think in both cases the error check should be moved to the top of the functions since elmt_size shouldn't change.

jhendersonHDF commented:
> I noticed you changed the code in realloc to always save need_insert, but it seems these allgathers are still here?

The function is still fundamentally the same, I just cleaned it up a bit to not need the inner loop anymore by having one path for the alloc call and then deciding whether the rank needs to update a local chunk. The need_insert status still gets thrown away by ranks that aren't updating a local chunk. I didn't do any work on the optimization we discussed for saving need_insert, so both allgathers are still taking place.

jhendersonHDF committed:
Simplify H5D__get_space_status for chunked datasets

H5D__get_num_chunks can be used to correctly determine space allocation
status for filtered and unfiltered chunked datasets

jhendersonHDF committed:
Return error for too many shared chunks during parallel filtered write

jhendersonHDF committed:
Move one-time check in H5D_select_io(_mem) outside of inner loop

jhendersonHDF committed:
Add notes for later parallel compression optimizations noted from review

brtnfld commented:
Performance testing in progress. Is there an issue I can attach results to?

brtnfld commented:
Using minIO--unstructured with gzip, on Cori, the memory issue seems to be much improved:

  _srun -n 1024 ./unstruct --pointspertask 16384 --tsteps 2 --hdf5 --hdf5_chunk 1024 961 9610  --hdf5_compress gzip,6_

HDF5-develop: Crashes 
HDF5-parallel-comp: Completes

jhendersonHDF commented:
> Using minIO--unstructured with gzip, on Cori, the memory issue seems to be much improved:
> 
> _srun -n 1024 ./unstruct --pointspertask 16384 --tsteps 2 --hdf5 --hdf5_chunk 1024 961 9610 --hdf5_compress gzip,6_
> 
> HDF5-develop: Crashes HDF5-parallel-comp: Completes

Good to hear! In just my testing on a local machine at small scale, writes appear to perform better and reads perform a little worse. I figured out why though and am deciding what to do there. Based on my profiling though, I imagine at scale the case I looked at won't have the same worse read performance characteristics.

jhendersonHDF commented:
> Performance testing in progress. Is there an issue I can attach results to?

Let's use TRILAB-283 for this.

brtnfld commented:
miniIO just does writing, I would need to use CGNS for both writing and reading.

jhendersonHDF committed:
Delay init of fill buffer until first unallocated chunk is found

jhendersonHDF commented:
Synced with develop for performance testing and added optimization to delay initialization of fill value buffer until first allocated chunk is found.

jhendersonHDF committed:
Parallel Compression performance optimization

Avoid doing I/O when a rank has no selection and the MPI communicator
size is 1 or the I/O has been requested as independent at the low level

Avoid 0-byte collective read of incrementally allocated filtered dataset
when dataset han't been written to yet

jhendersonHDF committed:
Add RELEASE.txt note for parallel compression improvements
